kubeadm: Add kubelet instance configuration to configure CRI socket for each node (#4658)
kubernetes/enhancements: keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> The proposal is to remove the Container Runtime Interface (CRI) socket annotation  named `kubeadm.alpha.kubernetes.io/cri-socket` from the Node object in Kubernetes, which is currently added during the `init upload-config` phase in Kubeadm. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 

Reviewer: ```suggestion
The proposal is to remove the Container Runtime Interface (CRI) socket annotation  named `kubeadm.alpha.kubernetes.io/cri-socket` from the Kubernetes Node object, which is currently added during the `kubeadm init upload-config` phase. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 
```

> instead of relying on the annotation, this KEP proses creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm init and join. This will eliminate the need for kubeadm to store CRI socket configuration on each Node object, instead relying on the kubeadm configuration file to specify this information.

Reviewer: ```suggestion
Instead of relying on the annotation, this KEP proses creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm commands. This will eliminate the need for kubeadm to store CRI socket configuration on each Node object.
```

## Motivation

Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.

After migrating the container runtime endpoint flag to the instance configuration, we can use it
Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

### Goals

* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.
* Provide an instance configuration file named `/var/lib/kubelet/instace-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.

> * The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file

Reviewer: ```suggestion
* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file.
```

### Non-Goals

- Continue maintaining CRI socket paths on Node objects.

> ## Proposal

Reviewer: the document still does not explain how exactly instance-config.yaml will override config.yaml
i suggest we use patches with the kubeletconfiguration target. this will allow use to override any field.

see here https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/

but also the user can pass --paches (or via --config), so the precedence (from high to low) must be:
- user provided kubeletconfig patches with --patches
- instance-config.yaml (uses kubeletconfig patch target to patch config.yaml)
- global config.yaml

WDYT?

if you agree to use the patches approach with kubeletconfig target then the above information must be added under proposal.

Reviewer: this part should also start talking about the feature gate.

Author: I agree with your opinion, I will add relevant explanations.

Author: After carefully reviewing the code, I think we should not directly use the `kubeletconfiguration patch target` mode to implement this feature, which will actually increase the complexity of the implementation. Instead, we should refer to its implementation process and merge `instance-config.yaml` and `kubelet-config.yaml` in the form of calling [ApplyPatchesToTarget](https://github.com/kubernetes/kubernetes/blob/6c556cb9c22d3f4686cbea41e6d301ba889fd6a6/cmd/kubeadm/app/util/patches/patches.go#L160).

Reviewer: https://github.com/kubernetes/kubernetes/blob/6c556cb9c22d3f4686cbea41e6d301ba889fd6a6/cmd/kubeadm/app/util/patches/patches.go#L160 does require the target in there to be specified as `kubeletconfiguration` and matching a known target but i think i understand what you mean. and this can be discussed on the k/k PRs too.

kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instance-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime endpoint will be automatically detected, uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.

> kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.

Reviewer: ```suggestion
kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration (`kube-system/kubelet-config`). If the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
```

but here i'm confused, if the CRI socket is not specified in kubeadm config this happens, right?
- download the global config and put it in config.yaml
- autodetect the socket on this node
- write the detected socket to instance-config.yaml

the above comment in your text is not correct.

Author: Yes, this is a mistake, the steps you mentioned are more in line with our current process.

> kubeadm upgrade:  future versions of `kubeadm upgrade apply/node` will only check `/var/lib/kubelet/instace-config.yaml`.

Reviewer: ```suggestion
kubeadm upgrade: future versions of `kubeadm upgrade apply/node` will only check `/var/lib/kubelet/instace-config.yaml`.
```

### Risks and Mitigations

## Design Details

**Alpha(v1.31):** 

kubeadm init:

* No longer need to write the `--container-runtime-endpoint` to `/var/lib/kubelet/kubeadm-flags.env`.
* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.

> * If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instace-config.yaml` configuration file is generated based on it. If the CRI socket is not set, the container runtime endpoint is automatically detected uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.

Reviewer: mistake instace->instance

check the whole doc for this problem please.

Reviewer: > and `/var/lib/kubelet/instace-config.yaml` will be generated.

you mean that during init both the global config and the instance-config.yaml will have the auto-detected socket?

Author: > you mean that during init both the global config and the instance-config.yaml will have the auto-detected socket?

Yes, but I think this should be a mistake. We should not patch the CRI socket of any node to the global kubelet config.

Reviewer: sorry, i still don't understand.
i think kubeadm should never use the socker that is provided in the global kubelet config that is in the config map?

today the annotation is always used unless the user passes a --config or --cri-socket that overrides it, correct?
the /var/lib/kubelet/config.yaml CRI socket field is new, but it's one of those things that don't make a lot of sense as kubeadm supports different container runtimes on different nodes.

Author: You are right, sorry for the misunderstanding, I meant the process I described in the KEP was wrong.

kubeadm join:

* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.

> * If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not set, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.

Reviewer: please adjust this text similarly to the text for init

kubeadm reset:

> * There is no need to do anything, as `/var/lib/kubelet/instace-config.yaml` will be cleaned up with the deletion of `/var/lib/kubelet.`

Reviewer: something to note here, the file will be deleted but it must be deleted only after we use the socket from the file because that is how we are going to connect to the CRI API to delete running containers.

is that a problem?

Author: This is not a problem for us because we first set `resetData.criSocketPath` and then remove the `/var/lib/kubelet` directory

I will explain it in more detail

kubeadm upgrade apply:

* `kubeadm upgrade apply` will check the `--container-runtime-endpoint` args in the `/var/lib/kubelet/kubeadm-flags.env` file and generate `/var/lib/kubelet/instace-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

> kubeadm upgrade node:

Reviewer: for 1.31 kubeadm upgrade node the plan should be the same as upgrade apply.

**Beta(v1.32)**:

* `kubeadm upgrade apply/node`  will use `/var/lib/kubelet/instace-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

**GA(v1.34)**:

* `kubeadm upgrade apply/node` will use `/var/lib/kubelet/instace-config.yaml` and remove CRI annotations from Node object and remove `--container-runtime-endpoint` args in `/var/lib/kubelet/kubeadm-flags.env` file

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

At least the following kubeadm packages will require updates and new unit tests:

* `cmd/kubeadm/app/cmd/phases/init`

* `cmd/kubeadm/app/phases/kubelet`

* `cmd/kubeadm/app/phases/upgrade`

* `cmd/kubeadm/app/cmd/phases/join`

##### Integration tests

- N/A

##### e2e tests

* A new e2e test will be added by using the kinder tool.

> ### Graduation Criteria

Reviewer: so there is a problem with the KEP currently

feature gates have two states: TRUE and FALSE

but the "Design details" and "Graduation Criteria" sections are organized as we both using and not using a feature gate, which is confusing.

your doc should be organized like this:

- in "Design details" explain the new feature gate
- say in which releases it will be alpha, beta, GA
- explain when it will be enabled by default
- explain what happens when it's enabled and not enabled to the different kubeadm commands

Author: I made some changes, please check if this is what we expected.

> ### Graduation Criteria

Reviewer: this section Graduation Criteria must explain what is required for the feature to graduate to Beta and GA . there is no graduation to Alpha.

**Alpha**: Users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to enable the `NodeLocalCRISocket` feature gate before calling kubeadm upgrade apply, which will allow a `/var/lib/kubelet/instace-config.yaml` to be generated and overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` with it.

**Beta**: Users can modify `ClusterConfiguration` to disable the feature gate during upgrades. This will allow them to continue using the CRI socket annotation on nodes.

**GA**: Users can no longer patch ClusterConfiguration to opt out of the feature and it will be locked to be enabled by default.

### Version Skew Strategy

kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` through `/var/lib/kubelet/instace-config.yaml`, it will be supported on all versions of the kubelet within the skew.

## Implementation History

- 2024-05-23: Initial draft KEP

## Drawbacks

> * This KEP will bring a breaking change. Many potential users (I believe it is the majority) still rely on the `kubeadm.alpha.kubernetes.io/cri-socket` annotation or the `/var/lib/kubelet/kubeadm-flags.env` file to declare the CRI socket endpoint on the Node, because many users are familiar with them.

Reviewer: most users don't really "rely" on it as it's an implementation detail.
some users do read / write it an "rely" on it, yes.

Instead of relying on the annotation, this KEP proses creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm commands. This will eliminate the need for kubeadm to store CRI socket configuration on each Node object.

## Motivation

Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.

After migrating the container runtime endpoint flag to the instance configuration, we can use it
Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

### Goals

* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.

> * Provide an instance configuration file named `/var/lib/kubelet/instance-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.

Reviewer 2: Just a naming question. 
We have another file in the same dir named `kubeadm-flags.env`. Should this fix be named with kubeadm prefix?

Reviewer: config.yaml is also kubeadm generated but is not prefixed with "kubeadm" 🤔
because of that i'm leaning towards using "instance-config.yaml", but happy to discuss more and hear more ideas around the naming.

kubeadm reset:

* There is no need to do anything, according to the existing process, we get CRISocketPath before deleting the /var/lib/kubelet directory, and after deleting the `/var/lib/kubelet` directory, `/var/lib/kubelet/instance-config.yaml` will also be cleaned up.

kubeadm upgrade:

* `kubeadm upgrade node/apply` will check the `--container-runtime-endpoint` args in the `/var/lib/kubelet/kubeadm-flags.env` file and generate `/var/lib/kubelet/instance-config.yaml` based on them, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

**In the Beta phase, the feature gate is enabled by default. If feature gate is disabled, kubeadm subcommands will not be changed, when the feature gate is enabled, the kubeadm subcommands change as follows:** 

* `kubeadm upgrade apply/node`  will use `/var/lib/kubelet/instance-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

> **In the Beta phase, the feature gate is enabled by default and cannot be disabled. the kubeadm subcommands change as follows:** 

Reviewer: ```suggestion
**In the GA phase, the feature gate is enabled by default and cannot be disabled. the kubeadm subcommands change as follows:** 
```

Reviewer: this was not updated?

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

At least the following kubeadm packages will require updates and new unit tests:

* `cmd/kubeadm/app/cmd/phases/init`

* `cmd/kubeadm/app/phases/kubelet`

* `cmd/kubeadm/app/phases/upgrade`

* `cmd/kubeadm/app/cmd/phases/join`

##### Integration tests

- N/A

##### e2e tests

* A new e2e test will be added by using the kinder tool.

### Graduation Criteria

#### Alpha

- Use `NodeLocalCRISocket` feature gate to implement features.
- Add corresponding e2e tests.
- Added documentation for feature gates.

#### Beta

* Make feature gate to be enabled by default.

- Gather feedback from developers and surveys.
- Update the feature gate documentation.
- Implement changes in kubeadm upgrade apply/node Beta phase.

#### GA

- Gather feedback from developers and surveys.
- Implement changes in kubeadm upgrade apply/node GA phase.

> - Update the phases documentation.

Reviewer 2: After GA, we may not add the old alpha annotation.
- We may deprecate the annotation in Beta?

We can keep the annotation if the node is upgraded from older version, and it can be removed manually if they want then.

Author: > We can keep the annotation if the node is upgraded from older version, and it can be removed manually if they want then.

Yes, this is what we want to do, once the feature gate is enabled, we will not set the CRI socket annotation during `kubeadm init/join`, nor remove it on upgrade.

So I think it has nothing to do with which phase we are in, because we already have feature gates to help us control version skew.

Reviewer: please add entry in the above list to remove the annotation from here too:
https://kubernetes.io/docs/reference/labels-annotations-taints/#kubeadm-alpha-kubernetes-io-cri-socket
once the FG goes GA.

this page also needs an update:
https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/
to tell the user to update instance-config instead of the annotation.

Reviewer: this was not updated.

Author: I appended them in subsequent lines.

### Upgrade / Downgrade Strategy

**Alpha**: Users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to enable the `NodeLocalCRISocket` feature gate before calling kubeadm upgrade apply, which will allow a `/var/lib/kubelet/instance-config.yaml` to be generated and overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` with it.

**Beta**: Users can modify `ClusterConfiguration` to disable the feature gate during upgrades. This will allow them to continue using the CRI socket annotation on nodes.

**GA**: Users can no longer patch ClusterConfiguration to opt out of the feature and it will be locked to be enabled by default.

### Version Skew Strategy

kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` through `/var/lib/kubelet/instance-config.yaml`, it will be supported on all versions of the kubelet within the skew.

## Implementation History

> - 2024-05-23: Initial draft KEP

Reviewer: ```suggestion
- 2024-05-23: Initial draft KEP
- <some-date>: KEP marked as implementable
```

After migrating the container runtime endpoint flag to the instance configuration, we can use it
Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

### Goals

* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and no longer set it.
* Provide an instance configuration file named `/var/lib/kubelet/instance-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.
* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file.

### Non-Goals

- Continue maintaining CRI socket paths on Node objects.

## Proposal

> We will add a new file `/var/lib/kubelet/instance-config.yaml` to customize the CRI socket of each node. This file will be merged with `/var/lib/kubelet/config.yaml` in the process of kubeadm init/join in a similar way to `kubeletconfiguration patch target` (refer to its code implementation), If the user uses the `--patch` args, the patch file provided by the user will be given priority.

Reviewer: ```suggestion
We will add a new file `/var/lib/kubelet/instance-config.yaml` to customize the CRI socket of each node. This file will be merged with `/var/lib/kubelet/config.yaml` in the process of kubeadm init/join in a similar way to `kubeletconfiguration` patch target (refer to its code implementation), If the user uses the `--patch` args, the patch file provided by the user will be given priority.
```

minor typo.
